### PR TITLE
fix(ci): Run each perf test in its own matrix job

### PR DESCRIPTION
**Pull request by jamilbk**
The iperf3 server sometimes hangs, or takes a while to startup.

Rather than trying to reset the iperf3 state between performance tests, this PR refactors them so they each run in their matrix job. This ensures each performance test will run on a separate VM, unaffected by previous test runs to eliminate the effect any residual network buffer state can have on a particular test.

It also makes sure the server is listening with a `healthcheck`.

### CODE REVIEW

**Review by thomaseizinger — APPROVED** [review]
I am not entirely convinced the bitrate is the problem but it is worth a try.

You could create a big matrix and do a couple of runs with 100s of tests to confirm.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -242,8 +242,1000 @@ jobs:
       matrix:
```
Comment (thomaseizinger) [inline .github/workflows/ci.yml:242]: This is normally how I do it. Does that not work?

```suggestion
      matrix:
      	dim1: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
      	dim2: ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']
```

Comment (jamilbk) [reply .github/workflows/ci.yml:242]: Hmm that seems to skip the `direct` test. Even copying them brute force seems to cause failures. Wondering now if it's something screwy with GH actions. Very strange

Comment (jamilbk) [reply .github/workflows/ci.yml:242]: Ah, need to fix main artifact name

**Comment by jamilbk** [comment]
Draft because I'm still testing in a matrix job to catch flakes.

File: scripts/tests/perf/setup.sh [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+
+set -euox pipefail
+
+# Establish a channel first. Helps the iperf3 test to be more stable.
```
Comment (thomaseizinger) [inline scripts/tests/perf/setup.sh]: Nit: I think "channel" is not clear. Are you talking about TURN channels? In connlib, we call this a "connection". "tunnel" would also make sense (one tunnel has many connections, some of which may use TURN channels via one or more relays).

File: scripts/tests/perf/relayed-udp-server2client.sh [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+
+set -euox pipefail
+
+source "./scripts/tests/perf/force-relayed.sh"
+source "./scripts/tests/perf/setup.sh"
+
+sleep 5
+docker compose exec --env RUST_LOG=info -it client /bin/sh -c 'iperf3 --cport 39999 --connect-timeout 5000 -t 30 -l 8K -M 1240 -Z -u -b 1M -R -c 172.20.0.110 --json' >>"${TEST_NAME}.json"
```
Comment (thomaseizinger) [inline scripts/tests/perf/relayed-udp-server2client.sh]: Can we use the long form for all options? Makes it a bit easier to understand.

File: scripts/tests/perf/relayed-udp-server2client.sh [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+
+set -euox pipefail
+
+source "./scripts/tests/perf/force-relayed.sh"
```
Comment (thomaseizinger) [inline scripts/tests/perf/relayed-udp-server2client.sh]: You could reuse the existing "lib" here I think. Also sending the ping is the same, right?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -240,14 +240,17 @@ jobs:
...
-          - test_name: direct-perf
-            setup: echo 'Noop'
-          - test_name: relayed-perf
-            setup: |
-              #  Disallow traffic between gateway and client container
-              sudo iptables -I FORWARD 1 -s 172.28.0.100 -d 172.28.0.105 -j DROP
-              sudo iptables -I FORWARD 1 -s 172.28.0.105 -d 172.28.0.100 -j DROP
+        dim1: [1, 2, 3, 4, 5]
+        dim2: ['a', 'b', 'c', 'd', 'e']
+        test_name:
```
Comment (thomaseizinger) [inline .github/workflows/ci.yml:243]: This work!

Comment (thomaseizinger) [reply .github/workflows/ci.yml:243]: This could also be an array (see other tests).

yaml style checker needed? 😁

**Review by thomaseizinger — APPROVED** [review]
Great work! Some minor feedback :)

File: .github/workflows/ci.yml (hunk 5 above)
Comment (jamilbk) [reply .github/workflows/ci.yml:243]: Yeah it's working now that I've refactored things a little bit ;-).

We now run each performance test on its own.

File: scripts/tests/perf/relayed-udp-server2client.sh (hunk 4 above)
Comment (jamilbk) [reply scripts/tests/perf/relayed-udp-server2client.sh]: Good call. I removed the ping -- no longer necessary.

File: scripts/tests/perf/relayed-udp-server2client.sh (hunk 3 above)
Comment (jamilbk) [reply scripts/tests/perf/relayed-udp-server2client.sh]: Sure, I'll add some documentaion too.

File: docker-compose.yml [hunk 6]
```diff
@@ -193,6 +192,8 @@ services:
 
   iperf3:
     image: mlabbe/iperf3
+    healthcheck:
```
Comment (jamilbk) [inline docker-compose.yml:195]: This is ultimately the root cause. It looks like even though the server started up and accepted connections, it wasn't "fully up" for some reason.

Waiting for Docker to acknowledge that the process is actually listening fixed the issue.

A `sleep 1` before running the test or something would probably fix it too.

File: scripts/tests/perf/setup.sh (hunk 2 above)
Comment (jamilbk) [reply scripts/tests/perf/setup.sh]: Noted!

**Comment by jamilbk** [comment]
Matrix is [passing reliably](https://github.com/firezone/firezone/actions/runs/7980331867), so removing.

Note that the result comparison changes in this PR is untested -- we'll need to wait until this lands in `main` for those to run. I'll fix any issues there if they come up.

**Pull request merged**